src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/GenotypeLikelihoodCalculatorDRAGEN.java

```diff
@@ -0,0 +1,487 @@
+            }
+            if (localMaxLpspi > maxLpspi) {
+                maxLpspi = Math.max(maxLpspi, localMaxLpspi);
+                lpfApplied = logProbFAllele;
+            }
+        }
+
+        //TODO javaize this
+        // TODO soon should not need to use the LPF applied here...
+        return new double[]{maxLpspi, lpfApplied};
+    }
+
+
+    // TODO: for reviewer... this code is meant to handle the performance regression brought about by FRD. Essentially in DRAGEN for every
+    // TODO  combination of strandedness and mapping quality a computation is done, this is vaguely unnecessary. Because this leads to
+    // TODO  a bias towards strandedness/allele combinations that have very few reads supporting them by virtue of the fact that a different
+    // TODO  strand/allele combination had at least one low mapping quality read. I have reverted the change right now and consequently this genotyping
+    // TODO  is taking somewhere in the order of ~5-6% runtime on the profiler whereas otherwise it could correspond to much less at the expense
+    // TODO  of not matching DRAGEN properly.
+    // helper method to populate the reads containers properly with their critical values and store them in the provided set
+    // NOTE: this has the side effect of setting the DragenReadContainer setPhredPFValue() values for the reads for the given set of alleles
+    private void computeCriticalValues(final Set<Double> criticalThresholdsForwards, final Set<Double> criticalThresholdsReverse, final Set<Double> criticalThresholdsTotal, final List<DRAGENGenotypesModel.DragenReadContainer> container, final double log10MapqPriorAdjustment) {
+        for (int i = 0; i < container.size(); i++) {
+            final DRAGENGenotypesModel.DragenReadContainer readContainer = container.get(i);
+            final double log10CriticalValue = readContainer.getPhredScaledMappingQuality() / -10.0 + log10MapqPriorAdjustment;
+            readContainer.setPhredPFValue(log10CriticalValue);
+            // Split the critical thresholds up by their applicable strands in order to avoid repeated work
+            if (readContainer.isReverseStrand()) {
+                criticalThresholdsReverse.add(log10CriticalValue);
+            } else {
+                criticalThresholdsForwards.add(log10CriticalValue);
+            }
+            criticalThresholdsTotal.add(log10CriticalValue);
+        }
+    }
+
+
+    /**
+     * See {@link GenotypeLikelihoodCalculator#genotypeLikelihoods}. This wrapper just enforces that the likelihoods object is recorded in the cache.
+     *
+     * @return never {@code null}.
+     */
+    public <EVIDENCE, A extends Allele> GenotypeLikelihoods genotypeLikelihoods(final LikelihoodMatrix<EVIDENCE, A> likelihoods) {
+        cachedLikelihoods = null;
```

Comment:
technically correct yes. all of this behavior is designed to very strictly enforce that we aren't possibly trying to genotype with the wrong set of read likelihoods (which is easy to do since everything lives in a series of flattened arrays). I will leave this by virtue of the fact that I want it to be eminenetly clear to future generations.